File: src/gallery-favorites.ts
import type {OnlineGalleryDetail, OnlineGalleryItem, OnlineGallerySourceId} from './online-gallery';
export const GALLERY_FAVORITES_KEY='online-gallery-favorites.v1';
export interface GalleryFavorite {source:OnlineGallerySourceId;id:string;title:string;author:string;sourceUrl:string;prompt:string;negativePrompt:string;createdAt:string;score:number;images:{url:string;thumb:string}[];savedAt:number;}
export type FavoriteSort='newest'|'oldest'|'name'|'name-desc'|'author'|'source'|'score';
export const favoriteKey=(item:Pick<GalleryFavorite,'source'|'id'>)=>`${item.source}:${item.id}`;
export function validGalleryUrl(value:string){try {return ['https:','http:'].includes(new URL(value).protocol)?value:'';}catch{return '';}}
export function parseFavorites(raw:string|null):GalleryFavorite[]{
 if(!raw)return [];const data=JSON.parse(raw);if(data.version!==1||!Array.isArray(data.items))throw new Error('Invalid favorites library');
 const seen=new Set<string>();return data.items.filter((x:GalleryFavorite)=>{
  if(!x||!['aitag','artist-ranking','danbooru','safebooru','gelbooru','quicktag','tags-gallery'].includes(x.source)||typeof x.id!=='string'||!x.id||typeof x.title!=='string'||!Array.isArray(x.images)||!Number.isFinite(x.savedAt))throw new Error('Invalid favorite');
  const key=favoriteKey(x);if(seen.has(key))return false;seen.add(key);return true;
 }).map((x:GalleryFavorite)=>({...x,author:String(x.author??''),prompt:String(x.prompt??''),negativePrompt:String(x.negativePrompt??''),createdAt:String(x.createdAt??''),score:Number(x.score)||0,sourceUrl:validGalleryUrl(x.sourceUrl),images:x.images.map(i=>({url:validGalleryUrl(i.url),thumb:validGalleryUrl(i.thumb)})).filter(i=>i.url)}));
}
export function changeFavorite(items:GalleryFavorite[],item:GalleryFavorite):GalleryFavorite[]{return items.some(x=>favoriteKey(x)===favoriteKey(item))?items.filter(x=>favoriteKey(x)!==favoriteKey(item)):[item,...items];}
export function orderFavorites(items:GalleryFavorite[],sort:FavoriteSort,language='en-US'){
 const name=new Intl.Collator(language,{numeric:true,sensitivity:'base'});return [...items].sort((a,b)=>{
 let n=sort==='oldest'?a.savedAt-b.savedAt:sort==='name'?name.compare(a.title,b.title):sort==='name-desc'?name.compare(b.title,a.title):sort==='author'?name.compare(a.author,b.author):sort==='source'?name.compare(a.source,b.source):sort==='score'?b.score-a.score:b.savedAt-a.savedAt;
 return n||favoriteKey(a).localeCompare(favoriteKey(b));});
}
export function favoriteFromGallery(item:OnlineGalleryItem,detail?:OnlineGalleryDetail):GalleryFavorite{return {source:item.source,id:item.id,title:item.title,author:item.author,sourceUrl:item.sourceUrl,prompt:detail?.prompt??item.prompt,negativePrompt:detail?.negativePrompt??item.negativePrompt,score:item.score,createdAt:item.createdAt,savedAt:Date.now(),images:(detail?.media.length?detail.media:[item.cover]).map(m=>({url:m.downloadUrl||m.displayUrl,thumb:m.previewUrl||m.displayUrl})).filter(m=>m.url)};}

File: mobile/lib/services/gallery_favorites.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'online_gallery_service.dart';

const galleryFavoritesKey = 'online-gallery-favorites.v1';
String validGalleryUrl(String value) {
  final u = Uri.tryParse(value);
  return u != null && ['http', 'https'].contains(u.scheme) && u.host.isNotEmpty
      ? value
      : '';
}

class GalleryFavorite {
  final String source,
      id,
      title,
      author,
      sourceUrl,
      prompt,
      negativePrompt,
      createdAt;
  final int score, savedAt;
  final List<Map<String, String>> images;
  const GalleryFavorite(
      {required this.source,
      required this.id,
      required this.title,
      this.author = '',
      this.sourceUrl = '',
      this.prompt = '',
      this.negativePrompt = '',
      this.createdAt = '',
      this.score = 0,
      required this.savedAt,
      this.images = const []});
  String get key => '$source:$id';
  Map<String, dynamic> toJson() => {
        'source': source,
        'id': id,
        'title': title,
        'author': author,
        'sourceUrl': sourceUrl,
        'prompt': prompt,
        'negativePrompt': negativePrompt,
        'createdAt': createdAt,
        'score': score,
        'savedAt': savedAt,
        'images': images
      };
  factory GalleryFavorite.fromJson(Map<String, dynamic> j) {
    if (![
          'aitag',
          'artist-ranking',
          'danbooru',
          'safebooru',
          'gelbooru',
          'quicktag',
          'tags-gallery'
        ].contains(j['source']) ||
        j['id'] is! String ||
        j['id'] == '' ||
        j['title'] is! String ||
        j['images'] is! List ||
        j['savedAt'] is! num) throw const FormatException('Invalid favorite');
    return GalleryFavorite(
        source: j['source'],
        id: j['id'],
        title: j['title'],
        author: '${j['author'] ?? ''}',
        sourceUrl: validGalleryUrl('${j['sourceUrl'] ?? ''}'),
        prompt: '${j['prompt'] ?? ''}',
        negativePrompt: '${j['negativePrompt'] ?? ''}',
        createdAt: '${j['createdAt'] ?? ''}',
        score: (j['score'] as num? ?? 0).toInt(),
        savedAt: (j['savedAt'] as num).toInt(),
        images: (j['images'] as List)
            .map((i) => <String, String>{
                  'url': validGalleryUrl('${i['url'] ?? ''}'),
                  'thumb': validGalleryUrl('${i['thumb'] ?? ''}')
                })
            .where((i) => i['url']!.isNotEmpty)
            .toList());
  }
  factory GalleryFavorite.fromItem(OnlineGalleryItem item,
          {OnlineGalleryDetail? detail}) =>
      GalleryFavorite(
          source: item.source.id,
          id: item.id,
          title: item.title,
          author: item.author,
          sourceUrl: item.sourceUrl,
          prompt: detail?.prompt ?? item.prompt,
          negativePrompt: detail?.negativePrompt ?? item.negativePrompt,
          createdAt: item.createdAt,
          score: item.score,
          savedAt: DateTime.now().millisecondsSinceEpoch,
          images: ((detail?.media.isNotEmpty ?? false)
                  ? detail!.media
                  : [item.cover])
              .map((m) => <String, String>{
                    'url':
                        m.downloadUrl.isNotEmpty ? m.downloadUrl : m.displayUrl,
                    'thumb':
                        m.previewUrl.isNotEmpty ? m.previewUrl : m.displayUrl
                  })
              .where((m) => m['url']!.isNotEmpty)
              .toList());
}

List<GalleryFavorite> parseFavorites(String? raw) {
  if (raw == null) return [];
  final j = jsonDecode(raw);
  if (j['version'] != 1 || j['items'] is! List) {
    throw const FormatException('Invalid favorites library');
  }
  final seen = <String>{};
  return (j['items'] as List)
      .map((i) => GalleryFavorite.fromJson(Map<String, dynamic>.from(i)))
      .where((i) => seen.add(i.key))
      .toList();
}

List<GalleryFavorite> orderFavorites(List<GalleryFavorite> items, String sort) {
  final out = [...items];
  out.sort((a, b) {
    final n = switch (sort) {
      'oldest' => a.savedAt.compareTo(b.savedAt),
      'name' => a.title.toLowerCase().compareTo(b.title.toLowerCase()),
      'name-desc' => b.title.toLowerCase().compareTo(a.title.toLowerCase()),
      'author' => a.author.toLowerCase().compareTo(b.author.toLowerCase()),
      'source' => a.source.compareTo(b.source),
      'score' => b.score.compareTo(a.score),
      _ => b.savedAt.compareTo(a.savedAt)
    };
    return n == 0 ? a.key.compareTo(b.key) : n;
  });
  return out;
}

class GalleryFavoritesStore extends ChangeNotifier {
  static final instance = GalleryFavoritesStore();
  List<GalleryFavorite> items = [];
  Object? error;
  Future<void>? _load;
  Future<void> _writes = Future.value();
  Future<void> load() => _load ??= () async {
        try {
          items = parseFavorites((await SharedPreferences.getInstance())
              .getString(galleryFavoritesKey));
          error = null;
        } catch (e) {
          error = e;
        }
        notifyListeners();
      }();
  Future<void> toggle(GalleryFavorite item) {
    final result = _writes.catchError((_) {}).then((_) async {
      await load();
      if (error != null) throw error!;
      final next = items.any((i) => i.key == item.key)
          ? items.where((i) => i.key != item.key).toList()
          : [item, ...items];
      final ok = await (await SharedPreferences.getInstance()).setString(
          galleryFavoritesKey,
          jsonEncode(
              {'version': 1, 'items': next.map((i) => i.toJson()).toList()}));
      if (!ok) throw StateError('Favorites save failed');
      items = next;
      notifyListeners();
    });
    _writes = result;
    return result;
  }
}
